File: c/src/functions/hyperoperations.c
/*
1. we must not have add_i(), add_f(), add_d(), add_cd(). we must have add(), which will dispatch the correct function
2. add(a, b) must never allow implicit type casting anywhere. add(2, 3.0) is strictly forbidden. there is absolutely no implicit type casting done anywhere

---------------

if i define a canonical data type for each number type, then the number type is restricted to that only. user wont be able to add two real floats. only two real doubles. this is bad..
if i define a canonical number type for each data type, then the operator knows that float is real, double is real, etc etc. this relation fits better. because many data types <-> one number type, but not many number type <-> one data type. 
so instead of defining add_integer, add_real, add_complex, i should have add_int, add_float, add_double, add_complex_float, … 

do not have typedefs like:

#define integer int
#define real double

do NOT do this!!! you will want to expand later. just make the operators for int, double, and double complex for now. oh and bool of course. but just do that for now!!! and later youll expand to more datatypes

----------------

operators are fully closed in these domains:
inc : NZRC
dec :  ZRC
add : NZRC
sub :  ZRC
mul : NZRC
div :   RC
pow : N  C
root:    C
log :    C

unsigned int is N
int          is Z
double       is R
complex      is C

later, we shall support float, uint* types, int*, float*, complex* etc etc. but lets keep these four for now, just to get a feel for the domains first
*/

#include "hyperoperations.h"
#include <errno.h>	// for error codes like EDOM
#include <math.h>


// incrementation. closed on N, Z, R, C
unsigned int _dm_inc_unsigned_int(unsigned int b)
{
	return ++b;
}
int _dm_inc_int (int b) 
{	
	return ++b;
}
double _dm_inc_double(double b)
{	
	return b + 1;
}
double complex _dm_inc_double_complex(double complex b) 
{	
	return b + 1;
}

// decrementation. closed on ℤ, ℝ, ℂ but not on ℕ
unsigned int _dm_dec_unsigned_int(unsigned int c) 
{	
	// since -1 doesnt exist for natural numbers ℕ
	if (c == 0)
		errno = EDOM;

	return --c;
}
int _dm_dec_int(int c)
{
	return --c;
}
double _dm_dec_double(double c)
{	
	return c - 1;
}
double complex _dm_dec_double_complex(double complex c)
{	
	return c - 1;
}

// addition. closed on N, Z, R, C
unsigned int _dm_add_unsigned_int(unsigned int a, unsigned int b)
{	
	return a + b;
}
int _dm_add_int(int a, int b)
{	
	return a + b;
}
double _dm_add_double(double a, double b) 
{	
	return a + b;
}
double complex _dm_add_double_complex(double complex a, double complex b)
{	
	return a + b;
}

// subtraction. closed on Z, R, C but not on N
unsigned int   _dm_sub_unsigned_int  (unsigned int   c, unsigned int   a)
{	
	// since c - a cannot be < 0
	if (c < a)
		errno = EDOM;
	
	return c - a;
}
int            _dm_sub_int           (int            c, int            a) {return c - a            ;}
double         _dm_sub_double        (double         c, double         a) {return c - a            ;}
double complex _sub_double_complex(double complex c, double complex a) {return c - a            ;}

// multiplication. closed on N, Z, R, C
unsigned int   _dm_mul_unsigned_int  (unsigned int   a, unsigned int   b) {return a * b            ;}
int            _dm_mul_int           (int            a, int            b) {return a * b            ;}
double         _dm_mul_double        (double         a, double         b) {return a * b            ;}
double complex _dm_mul_double_complex(double complex a, double complex b) {return a * b            ;}

// division. closed on R, C only
unsigned int _dm_div_unsigned_int(unsigned int c, unsigned int a)
{	
	if (c % a != 0)
		errno = EDOM;

	return c / a;
}
int _dm_div_int (int c, int a)
{	
	if (c % a != 0)
		errno = EDOM;

	return c / a;
}
double _dm_div_double(double c, double a)
{	
	return c / a;
}
double complex _dm_div_double_complex(double complex c, double complex a)
{	
	return c / a;
}

// exponentiation. closed on N, C only
unsigned int _dm_pow_unsigned_int(unsigned int a, unsigned int b)
{
	unsigned int result = 1;
	
	while (b > 0)
	{	
		if (b & 1)	// if current bit of exponent is 1
			result *= a;	// multiply result by a

		a *= a;	// square the base
		b >>= 1;	// shift exponent right by 1 (divide by 2)
	}
	
	return result;
}
int _dm_pow_int(int a, int b)
{
	/*              b
	 *   -27 -8 -1  0  1  8 27 
	 *     9  4  1  0  1  4  9
	 *    -3 -2 -1  0  1  2  3   
	 * -a  1  1  1  ?  1  1  1  a
	 *     X  X -1  0  1  X  X
	 *     X  X  1  0  1  X  X
	 *     X  X -1  0  1  X  X
	 *             -b
	 */
	
	if ((b < 0) && ((a == -1) || (a == 1)))
		errno = EDOM;

	if ((a == 0) && (b == 0))
		errno = EDOM;
	
	return pow(a, b);
}
int dm_dm_pow_i(int a, int b)
{
    if (a == 0 && b == 0) { errno = EDOM; return 0; }
    if (a == 0)           { return 0; }
    if (b < 0)            { errno = EDOM; return 0; } // not closed on ℤ unless |a|==1
    
    int result = 1;
    while (b > 0)
    {
        if (b % 2 == 1)
            result *= a;
        a *= a;
        b /= 2;
    }
    return result;
}
double _dm_pow_double(double a, double b)
{
	return pow(a, b);
}
double complex _dm_pow_double_complex(double complex a, double complex b)
{	
	return cpow(a, b);
}

// logarithm. closed on C only
double complex _dm_log_double_complex(double complex a, double complex c) {return clog(c) / clog(a);}

// root. closed on C only
double _dm_root_double(double b, double c)
{	
	return cpow(c, 1.0 / b) ;
}
double complex _dm_root_double_complex(double complex b, double complex c) {return cpow(c, 1.0 / b) ;}

/*
* daamath supports everything up to complex numbers because complex numbers are algebraically closed. that is very nice
* but the tradeoff is that complex algebra often has branches (multiple outputs for one input) and we want functions to give only one output. so the tradeoff that daamath makes is that it picks conventional principal branches

* in fact, i want to support all geometric algebras by making an algebra generator like ganja.js :( .....
* but i dont have the mathematical expertise yet. one day... GA generator + hyperoperation generator. it would be amazing.. plus, we have a string library too! and constants. etc etc. i love this library
*/

File: implementations/python/src/daamath/functions/hyperoperations.py
import daamath as dm
from ..exceptions import DomainViolation as _DomainViolation, CodomainViolation as _CodomainViolation

def succ(a):
    'successor. solves for b in ++a = b'
    if not dm.context.succ.domains.a(a):
        raise _DomainViolation('succ', (a), {}, a, dm.context.succ.domains.a)

    image = dm.context.succ.mapping(a)

    if not dm.context.succ.codomain(image):
        raise _CodomainViolation('succ', (a), {}, image, dm.context.succ.codomain)
    return image

def pred(b):
    'predecessor. solves for a in ++a = b'
    if not dm.context.pred.domains.b(b):
        raise _DomainViolation('pred', (b), {}, b, dm.context.pred.domains.b)

    image = dm.context.pred.mapping(b)

    if not dm.context.pred.codomain(image):
        raise _CodomainViolation('pred', (b), {}, image, dm.context.pred.codomain)
    return image

def add(a, b):
    'addision. solves for c in a + b = c'
    if not dm.context.add.domains.a(a):
        raise _DomainViolation('add', (a, b), {}, a, dm.context.add.domains.a)
    if not dm.context.add.domains.b(b):
        raise _DomainViolation('add', (a, b), {}, b, dm.context.add.domains.b)

    image = dm.context.add.mapping(a, b)

    if not dm.context.add.codomain(image):
        raise _CodomainViolation('add', (a, b), {}, image, dm.context.add.codomain)
    return image

def sub(c, b):
    'subtraction. solves for a in a + b = c'
    if not dm.context.sub.domains.c(c):
        raise _DomainViolation('sub', (c, b), {}, c, dm.context.sub.domains.c)
    if not dm.context.sub.domains.b(b):
        raise _DomainViolation('sub', (c, b), {}, b, dm.context.sub.domains.b)

    image = dm.context.sub.mapping(c, b)

    if not dm.context.sub.codomain(image):
        raise _CodomainViolation('sub', (c, b), {}, image, dm.context.sub.codomain)
    return image

def bus(c, a):
    'bustraction. solves for b in a + b = c'
    if not dm.context.bus.domains.c(c):
        raise _DomainViolation('bus', (c, a), {}, c, dm.context.bus.domains.c)
    if not dm.context.bus.domains.a(a):
        raise _DomainViolation('bus', (c, a), {}, a, dm.context.bus.domains.a)

    image = dm.context.bus.mapping(c, a)

    if not dm.context.bus.codomain(image):
        raise _CodomainViolation('bus', (c, a), {}, image, dm.context.bus.codomain)
    return image

def mul(a, b):
    'multiplication. solves for c in a × b = c'
    if not dm.context.mul.domains.a(a):
        raise _DomainViolation('mul', (a, b), {}, a, dm.context.mul.domains.a)
    if not dm.context.mul.domains.b(b):
        raise _DomainViolation('mul', (a, b), {}, b, dm.context.mul.domains.b)

    image = dm.context.mul.mapping(a, b)

    if not dm.context.mul.codomain(image):
        raise _CodomainViolation('mul', (a, b), {}, image, dm.context.mul.codomain)
    return image

def div(c, b):
    'division. solves for a in a × b = c'
    if not dm.context.div.domains.c(c):
        raise _DomainViolation('div', (c, b), {}, c, dm.context.div.domains.c)
    if not dm.context.div.domains.b(b):
        raise _DomainViolation('div', (c, b), {}, b, dm.context.div.domains.b)

    image = dm.context.div.mapping(c, b)

    if not dm.context.div.codomain(image):
        raise _CodomainViolation('div', (c, b), {}, image, dm.context.div.codomain)
    return image

def vid(c, a):
    'vidision. solves for b in a × b = c'
    if not dm.context.vid.domains.c(c):
        raise _DomainViolation('vid', (c, a), {}, c, dm.context.vid.domains.c)
    if not dm.context.vid.domains.a(a):
        raise _DomainViolation('vid', (c, a), {}, a, dm.context.vid.domains.a)

    image = dm.context.vid.mapping(c, a)

    if not dm.context.vid.codomain(image):
        raise _CodomainViolation('vid', (c, a), {}, image, dm.context.vid.codomain)
    return image

def pow(a, b):
    'exponentiation. solves for c in a ^ b = c'
    if not dm.context.pow.domains.a(a):
        raise _DomainViolation('pow', (a, b), {}, a, dm.context.pow.domains.a)
    if not dm.context.pow.domains.b(b):
        raise _DomainViolation('pow', (a, b), {}, b, dm.context.pow.domains.b)

    image = dm.context.pow.mapping(a, b)

    if not dm.context.pow.codomain(image):
        raise _CodomainViolation('pow', (a, b), {}, image, dm.context.pow.codomain)
    return image

def log(c, b):
    'logarithm. solves for a in a ^ b = c'
    if not dm.context.log.domains.c(c):
        raise _DomainViolation('log', (c, b), {}, c, dm.context.log.domains.c)
    if not dm.context.log.domains.b(b):
        raise _DomainViolation('log', (c, b), {}, b, dm.context.log.domains.b)

    image = dm.context.log.mapping(c, b)

    if not dm.context.log.codomain(image):
        raise _CodomainViolation('log', (c, b), {}, image, dm.context.log.codomain)
    return image

def root(c, a):
    'root. solves for b in a ^ b = c'
    if not dm.context.root.domains.c(c):
        raise _DomainViolation('root', (c, a), {}, c, dm.context.root.domains.c)
    if not dm.context.root.domains.a(a):
        raise _DomainViolation('root', (c, a), {}, a, dm.context.root.domains.a)

    image = dm.context.root.mapping(c, a)

    if not dm.context.root.codomain(image):
        raise _CodomainViolation('root', (c, a), {}, image, dm.context.root.codomain)
    return image
